packages/components/src/form-token-field/suggestions-list.js

```diff
@@ -1,158 +1,150 @@
 /**
  * External dependencies
  */
-import { map } from 'lodash';
-import scrollIntoView from 'dom-scroll-into-view';
+import { map, noop } from 'lodash';
+import scrollView from 'dom-scroll-into-view';
 import classnames from 'classnames';
 
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
+import { useEffect, useState } from '@wordpress/element';
 import { withSafeTimeout } from '@wordpress/compose';
 
-class SuggestionsList extends Component {
-	constructor() {
-		super( ...arguments );
-		this.handleMouseDown = this.handleMouseDown.bind( this );
-		this.bindList = this.bindList.bind( this );
-	}
+const emptyList = Object.freeze( [] );
 
-	componentDidUpdate() {
+const handleMouseDown = ( e ) => {
+	// By preventing default here, we will not lose focus of <input> when clicking a suggestion
+	e.preventDefault();
+};
+
+function SuggestionsList( {
+	selectedIndex,
+	scrollIntoView,
+	match = '',
+	onHover = noop,
+	onSelect = noop,
+	suggestions = emptyList,
+	displayTransform,
+	instanceId,
+	setTimeout,
+} ) {
+	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
+	const [ list, setList ] = useState();
```

Comment:
Instead of using state and a custom ref callback to store a `ref` to the `<ul>` element, we can use the [custom `useRefEffect` hook](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-ref-effect/index.ts):

<details>
<summary>Click to expand suggested changes</summary>

```diff
diff --git a/packages/components/src/form-token-field/suggestions-list.js b/packages/components/src/form-token-field/suggestions-list.js
index 397a33863e..1c0d770af6 100644
--- a/packages/components/src/form-token-field/suggestions-list.js
+++ b/packages/components/src/form-token-field/suggestions-list.js
@@ -8,8 +8,8 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
-import { useEffect, useState } from '@wordpress/element';
-import { withSafeTimeout } from '@wordpress/compose';
+import { useState } from '@wordpress/element';
+import { withSafeTimeout, useRefEffect } from '@wordpress/compose';
 
 const emptyList = Object.freeze( [] );
 
@@ -30,30 +30,27 @@ function SuggestionsList( {
 	setTimeout,
 } ) {
 	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
-	const [ list, setList ] = useState();
+	const listRef = useRefEffect(
+		( listNode ) => {
+			// only have to worry about scrolling selected suggestion into view
+			// when already expanded
+			if (
+				selectedIndex > -1 &&
+				scrollIntoView &&
+				listNode.children[ selectedIndex ]
+			) {
+				setScrollingIntoView( true );
+				scrollView( listNode.children[ selectedIndex ], listNode, {
+					onlyScrollIfNeeded: true,
+				} );
 
-	useEffect( () => {
-		// only have to worry about scrolling selected suggestion into view
-		// when already expanded
-		if (
-			selectedIndex > -1 &&
-			scrollIntoView &&
-			list.children[ selectedIndex ]
-		) {
-			setScrollingIntoView( true );
-			scrollView( list.children[ selectedIndex ], list, {
-				onlyScrollIfNeeded: true,
-			} );
-
-			setTimeout( () => {
-				setScrollingIntoView( false );
-			}, 100 );
-		}
-	}, [ selectedIndex, scrollIntoView, list ] );
-
-	const bindList = ( ref ) => {
-		setList( ref );
-	};
+				setTimeout( () => {
+					setScrollingIntoView( false );
+				}, 100 );
+			}
+		},
+		[ selectedIndex, scrollIntoView ]
+	);
 
 	const handleHover = ( suggestion ) => {
 		return () => {
@@ -98,7 +95,7 @@ function SuggestionsList( {
 	// TODO does this still apply now that it's a <ul> and not a <div>?
 	return (
 		<ul
-			ref={ bindList }
+			ref={ listRef }
 			className="components-form-token-field__suggestions-list"
 			id={ `components-form-token-suggestions-${ instanceId }` }
 			role="listbox"
```
</details>

This could potentially help to fix the [e2e test error](https://github.com/WordPress/gutenberg/runs/4575862072?check_suite_focus=true), since the callback in `useRefEffect` is invoked only if `listNode` isn't falsy.

Reply:
Can you please help me in understanding where it was wrong in my code? Just curious

Reply:
There wasn't anything _wrong_ in your code — you were just refactoring the component from class to function — my suggestion here is more of an improvement on top of your changes.

Here's the reasoning that brought me to suggest `useRefEffect`:

- While reviewing your PR, I was having a look at the `list` state, and noticed that it was used to hold a `ref`.
- Now, we could have used `useRef` instead, but then I noticed that `list` was used as a dependency of the `useEffect` hook — switching to `useRef` would meant that changes to `list` wouldn't trigger the `useEffect` hook.
- The proper way to refactor the code, then, would be to rewrite `list` to use `useCallback` to hold the `ref`. The code in `useEffect` (and its dependencies) could have moved to the body of `useCallback` too
- Finally, I realised that the pattern that I just described is implemented in a custom hook called [`useRefEffect`](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-ref-effect/index.ts)

Reply:
Thanks for the brief explanation.